service/src/main/kotlin/fi/espoo/evaka/koski/KoskiData.kt

```diff
@@ -237,30 +275,31 @@ data class KoskiActiveDataRaw(
  * `inclusiveRanges`: a sequence of non-overlapping date ranges that are considered *inclusive* (child is present)
  * `clampRange`: an optional clamping range. If available, all returned date ranges are guaranteed to be contained within this clamping range
  */
-fun calculateStudyRightRanges(
-    inclusiveRanges: Sequence<ClosedPeriod>,
-    clampRange: ClosedPeriod? = null
-): List<ClosedPeriod> {
-    val iter = inclusiveRanges.sortedWith(compareBy({ it.start }, { it.end }))
-        .mapNotNull { if (clampRange == null) it else clampRange.intersection(it) }
-        .iterator()
+internal fun calculateStudyRightTimelines(
+    placementRanges: Sequence<ClosedPeriod>,
+    holidays: Set<LocalDate>,
+    absences: Sequence<KoskiPreparatoryAbsence>
+): StudyRightTimelines {
+    val placement = Timeline().addAll(placementRanges)
+    val plannedAbsence = Timeline().addAll(
+        Timeline()
+            .addAll(absences.filter { it.type == AbsenceType.PLANNED_ABSENCE }.map { it.date.toClosedPeriod() })
+            .fillWeekendAndHolidayGaps(holidays)
+            .intersection(placement)
+            .periods().filter { it.durationInDays() > 7 }
+    )
+    val unknownAbsence = Timeline().addAll(
+        Timeline()
+            .addAll(absences.filter { it.type == AbsenceType.UNKNOWN_ABSENCE }.map { it.date.toClosedPeriod() })
+            .fillWeekendAndHolidayGaps(holidays)
+            .intersection(placement)
+            .periods().filter { it.durationInDays() > 7 }
+    )
 
-    var current: ClosedPeriod? = null
-    val result = mutableListOf<ClosedPeriod>()
-    while (iter.hasNext()) {
-        current = if (current == null) {
-            iter.next()
-        } else {
-            val next = iter.next()
-            check(current.end < next.start)
-            if (current.end.plusDays(1) == next.start) {
-                ClosedPeriod(current.start, next.end)
-            } else {
-                result.add(current)
-                next
-            }
-        }
-    }
-    current?.let { result.add(it) }
-    return result
+    return StudyRightTimelines(
+        placement = placement,
+        present = placement.removeAll(plannedAbsence).removeAll(unknownAbsence),
+        plannedAbsence = plannedAbsence,
+        unknownAbsence = unknownAbsence
+    ).also { println(it) }
```

Comment:
Was this for developing only or still needed?

Reply:
Oops :see_no_evil: